src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadAsyncState.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class ReadAsyncState : IDisposable
```

Comment:
Correct. Do you think we should reserve this type for IAE deserialization methods only?

Reply:
Curious - what state here is not needed for the non-IAE cases?

Reply:
I'm actually starting to think that `ReadAsyncState` is not needed at all at this point. I believe it was added to support `ContinueDeserialize` in the `IAsyncEnumerator`, however we [just refactored it using iterator methods](https://github.com/dotnet/runtime/pull/50778#pullrequestreview-630072247) so we can probably get rid of it.

Reply:
I ended up with the following change: https://github.com/dotnet/runtime/pull/50778/commits/c062017d639b9091d4e5e231caaf7e20e2de0ad0

Since `ReadAsyncState` contains the entire mutable state of async deserialization, it cannot be trivially converted into a struct. At the same time, lifting all of the state to the state machine bodies would result in duplication of the buffer management logic for both regular and IAE deserialization. 

What I did instead is create a struct that only contains buffer state and moved everything else to the state machines. This lets us keep the existing structure with the `ReadFromStreamAsync` and `ContinueDeserialize` methods being reused in both public async deserialization methods. Note that there is a slight amount of copying required in the case of `ReadFromStreamAsync` but the struct is small enough and I think this strikes the right trade-off between performance and DRY concerns.

Reply:
The new struct + the new iterator looks good.